context.go

```diff
@@ -0,0 +1,231 @@
+// functions for error and model marshaling in handler functions.
+type Context interface {
+	context.Context
+	http.ResponseWriter
+
+	// AddError adds a new error to the list of errors for this request.
+	AddError(err error)
+
+	// HasError returns true if at least one error has been added to the context.
+	HasError() bool
+
+	// WriteError writes out an HTTP status code, friendly error message, and
+	// optionally a set of error details set with `AddError` and/or passed in.
+	WriteError(status int, message string, errors ...error)
+
+	// WriteModel writes out an HTTP status code and marshalled model based on
+	// content negotiation (e.g. JSON or CBOR). This must match the registered
+	// response status code & type.
+	WriteModel(status int, model interface{})
+}
+
+type hcontext struct {
+	context.Context
+	http.ResponseWriter
+	r      *http.Request
+	errors []error
+	op     *Operation
+	closed bool
+}
+
+func (c *hcontext) AddError(err error) {
+	c.errors = append(c.errors, err)
+}
+
+func (c *hcontext) HasError() bool {
+	return len(c.errors) > 0
+}
+
+func (c *hcontext) WriteHeader(status int) {
+	if c.op != nil {
+		allowed := []string{}
+		for _, r := range c.op.responses {
+			if r.status == status {
+				for _, h := range r.headers {
+					allowed = append(allowed, h)
+				}
+			}
+		}
+
+		// Check that all headers were allowed to be sent.
+		for name := range c.Header() {
+			found := false
+
+			for _, h := range allowed {
+				if strings.ToLower(name) == strings.ToLower(h) {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				panic(fmt.Errorf("Header %s is not declared for %s %s (allowed: %s)", name, c.r.Method, c.r.URL.Path, allowed))
+			}
+		}
+	}
+
+	c.ResponseWriter.WriteHeader(status)
+}
+
+func (c *hcontext) Write(data []byte) (int, error) {
+	if c.closed {
+		panic(fmt.Errorf("Trying to write to response after WriteModel or WriteError for %s %s", c.r.Method, c.r.URL.Path))
+	}
+
+	return c.ResponseWriter.Write(data)
+}
+
+func (c *hcontext) WriteError(status int, message string, errors ...error) {
+	details := []*ErrorDetail{}
+
+	c.errors = append(c.errors, errors...)
+	for _, err := range c.errors {
+		if d, ok := err.(ErrorDetailer); ok {
+			details = append(details, d.ErrorDetail())
+		} else {
+			details = append(details, &ErrorDetail{Message: err.Error()})
+		}
+	}
+
+	model := &ErrorModel{
+		Title:  http.StatusText(status),
+		Status: status,
+		Detail: message,
+		Errors: details,
+	}
+
+	// Select content type and transform it to the appropriate error type.
+	ct := selectContentType(c.r)
+	switch ct {
+	case "application/cbor":
+		ct = "application/problem+cbor"
+	case "", "application/json":
+		ct = "application/problem+json"
+	case "application/yaml", "application/x-yaml":
+		ct = "application/problem+yaml"
+	}
+
+	c.writeModel(ct, status, model)
+}
+
+func (c *hcontext) WriteModel(status int, model interface{}) {
+	// Get the negotiated content type the client wants and we are willing to
```

Comment:
check that the context is closed just like in https://github.com/danielgtaylor/huma/pull/11/files#diff-05a3aa0ab9e5687ff167e0ef3ae4c5a6R97